File: nodejs/tests/gyroscope/l3g4200d_test.js
'use strict';
const { I2CConnection } = require('../../packages/periph/src/connection/i2c');
const { L3G4200DMinimal, L3G4200DFull } = require('../../packages/periph/src/chips/gyroscope/l3g4200d');

const I2C_BUS  = parseInt(process.env.I2C_BUS  || '1',  10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR || '0x68', 16);

let passed = 0, failed = 0;

async function main() {
    const connection = new I2CConnection(I2C_BUS, I2C_ADDR);

    const gyro = new L3G4200DMinimal(connection);
    if (gyro) { console.log('PASS init'); passed++; }
    else { console.log('FAIL init'); failed++; }

    const [x, y, z] = await gyro.angularRate();
    console.log(`PASS angular_rate x=${x.toFixed(3)} y=${y.toFixed(3)} z=${z.toFixed(3)} rad/s`);
    passed++;

    const gyroFull = new L3G4200DFull(connection);
    if (gyroFull) { console.log('PASS full_init'); passed++; }
    else { console.log('FAIL full_init'); failed++; }

    const cid = await gyroFull.whoAmI();
    if (cid === 0xD3) { console.log('PASS who_am_i'); passed++; }
    else { console.log('FAIL who_am_i'); failed++; }

    await gyroFull.configure(1, 0, 500);
    if (gyroFull._fullScale === 500) { console.log('PASS configure'); passed++; }
    else { console.log('FAIL configure'); failed++; }

    const status = await gyroFull.status();
    console.log(`PASS status 0x${status.toString(16)}`);
    passed++;

    const temp = await gyroFull.temperature();
    if (temp >= -50 && temp <= 100) { console.log('PASS temperature_range'); passed++; }
    else { console.log('FAIL temperature_range'); failed++; }

    await connection.close();
    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();

File: jvm/tests/gyroscope/l3g4200d/L3G4200DTest.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 22+
//JAVA_OPTIONS --enable-native-access=ALL-UNNAMED
//DEPS it.uhde:periph-connection:1.2.0
//DEPS it.uhde:periph-java:1.2.0

import it.uhde.periph.connection.I2CConnection;
import it.uhde.periph.chips.gyroscope.L3g4200dMinimal;
import it.uhde.periph.chips.gyroscope.L3g4200dFull;

public class L3G4200DTest {
    static int passed = 0, failed = 0;

    static void checkTrue(boolean cond, String label) {
        if (cond) { System.out.println("PASS " + label); passed++; }
        else      { System.out.println("FAIL " + label); failed++; }
    }

    public static void main(String[] args) throws Exception {
        int bus = Integer.parseInt(System.getenv().getOrDefault("I2C_BUS", "1"));
        int addr = Integer.parseInt(
            System.getenv().getOrDefault("I2C_ADDR", "0x68").replaceFirst("^0[xX]", ""), 16);

        try (var connection = new I2CConnection(bus, addr)) {
            var gyro = new L3g4200dMinimal(connection, false);
            float[] xyz = gyro.angularRate();
            checkTrue(xyz[0] >= -50.0f && xyz[0] <= 50.0f, "angular_rate_x_range");
            checkTrue(xyz[1] >= -50.0f && xyz[1] <= 50.0f, "angular_rate_y_range");
            checkTrue(xyz[2] >= -50.0f && xyz[2] <= 50.0f, "angular_rate_z_range");

            try (var connection2 = new I2CConnection(bus, addr)) {
                var gyroFull = new L3g4200dFull(connection2, false);
                checkTrue(gyroFull.whoAmI() == 0xD3, "who_am_i");
                gyroFull.configure(L3g4200dFull.ODR_200_HZ, 0, L3g4200dFull.FS_500_DPS);
                checkTrue(gyroFull.getClass().equals(L3g4200dFull.class), "configure");
                float[] xyz2 = gyroFull.angularRate();
                checkTrue(xyz2[0] >= -500.0f && xyz2[0] <= 500.0f, "configure_then_read");
                checkTrue(gyroFull.status() <= 0xFF, "status_readable");
                int temp = gyroFull.temperature();
                checkTrue(temp >= -50 && temp <= 100, "temperature_range");
                gyroFull.setFullScale(L3g4200dFull.FS_2000_DPS);
                float[] xyz3 = gyroFull.angularRate();
                checkTrue(xyz3[0] >= -2000.0f && xyz3[0] <= 2000.0f, "set_full_scale_2000");
                int samples = gyroFull.fifoSamples();
                checkTrue(samples <= 31, "fifo_samples_in_range");
                gyroFull.enableFifo(L3g4200dFull.FIFO_STREAM, 10);
                checkTrue(true, "enable_fifo_no_throw");
            }
        }
        System.out.printf("===DONE: %d passed, %d failed===%n", passed, failed);
        if (failed != 0) System.exit(1);
    }
}
